### PR TITLE
Make expansions stable for incr. comp.

### PR DESCRIPTION
**Pull request by cjgillot**
This PR aims to make expansions stable for incr. comp. by using the same architecture as definitions:
- the interned identifier `ExpnId` contains a `CrateNum` and a crate-local id;
- bidirectional maps `ExpnHash <-> ExpnId` are setup;
- incr. comp. on-disk cache saves and reconstructs expansions using their `ExpnHash`.

I tried to use as many `LocalExpnId` as I could in the resolver code, but I may have missed a few opportunities.

All this will allow to use an `ExpnId` as a query key, and to force this query without recomputing caller queries. For instance, this will be used to implement #85999.

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
(I'll get to this next weekend, most likely.)

**Comment by petrochenkov** [comment]
Phew, I did a first pass over this PR just to understand what's going on, and it took a ~whole working day because the PR does a lot, it would be better to break it into parts if possible.

**Comment by petrochenkov** [comment]
The fundamental difference between hashes used for definitions and expansions right now is that in case of definitions we a hashing something that is "resistant to small changes" (a def path), but in case of expansions we are just accumulating the whole `ExpnData` into the hash.
So if we, for example, move the location of a macro call slightly, its expansion hash will change as well.
But is doesn't really affects the infra (all the `ExpnHash <-> ExpnId` maps, etc), so we'll be able to change what exactly goes into the hash after landing this PR, that's good.

**Comment by petrochenkov** [comment]
General comments:
- I'd prefer to see a newtype a la `DefIndex` for local indices instead of a raw `u32`s.
- Commits like https://github.com/rust-lang/rust/pull/86676/commits/e3abfd12f2e744ad4de888ab439f71a15c62e02e need comments, otherwise it's not clear why we are introducing duplication / copypaste for no apparent reason.
- It would be good to avoid back and forth like changing `ExpnId` to `LocalExpnId` in one commit and then changing it back in some later commit.
- Style nits: please use `impl Trait` in the first commit touching code in which it's usable; please don't use `Self` outside of generics, it almost always makes readability worse.

I doubt that it's possible to extract some "pure refactoring" parts from this PR, because the whole PR is kind of a gradual refactoring, but if it's possible, then it would be great.

**Comment by cjgillot** [comment]
> Phew, I did a first pass over this PR just to understand what's going on, and it took a ~whole working day because the PR does a lot, it would be better to break it into parts if possible.

Sorry for the lack of explanations.

> The fundamental difference between hashes used for definitions and expansions right now is that in case of definitions we a hashing something that is "resistant to small changes" (a def path), but in case of expansions we are just accumulating the whole `ExpnData` into the hash.
> So if we, for example, move the location of a macro call slightly, its expansion hash will change as well.
> But it doesn't really affects the infra (all the `ExpnHash <-> ExpnId` maps, etc), so we'll be able to change what exactly goes into the hash after landing this PR, that's good.

I agree with this concern. I am trying to design an extension to #84373 which makes spans stable enough to resolve it.

> General comments:
> * I'd prefer to see a newtype a la `DefIndex` for local indices instead of a raw `u32`s.
> * Commits like e3abfd1 need comments, otherwise it's not clear why we are introducing duplication / copypaste for no apparent reason.
> * It would be good to avoid back and forth like changing  `ExpnId` to `LocalExpnId` in one commit and then changing it back in some later commit.
> * Style nits: please use `impl Trait` in the first commit touching code in which it's usable; please don't use `Self` outside of generics, it almost always makes readability worse.

I added a few commits and the current ones to avoid such back-and-forth. I could not find which functions you are referring to with `impl Trait`.

> I doubt that it's possible to extract some "pure refactoring" parts from this PR, because the whole PR is kind of a gradual refactoring, but if it's possible, then it would be great.

I extracted #87044 if it helps.

**Comment by petrochenkov** [comment]
>I could not find which functions you are referring to with impl Trait.

`fn decode_expn_id` and all similar functions taking `impl Fn(Mut,Once)` callbacks as arguments.

**Comment by petrochenkov** [comment]
Blocked on #87044.

**Comment by petrochenkov** [comment]
>I extracted #87044 if it helps.

It certainly helped.

**Comment by petrochenkov** [comment]
@cjgillot 
Do you plan to extract more parts, or this is ready for review?

**Comment by cjgillot** [comment]
@petrochenkov I don't have any more parts to extract. This is ready for review.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 1]
```diff
@@ -31,7 +31,7 @@ use rustc_session::config::CrateType;
 use rustc_span::symbol::{sym, Ident, Symbol};
 use rustc_span::{self, ExternalSource, FileName, SourceFile, Span, SyntaxContext};
 use rustc_span::{
-    hygiene::{ExpnDataEncodeMode, HygieneEncodeContext, MacroKind},
+    hygiene::{ExpnIndex, HygieneEncodeContext, MacroKind},
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:34]: Nit: could you avoid multi-line and nested imports?

File: compiler/rustc_middle/src/ty/query/on_disk_cache.rs [hunk 2]
```diff
@@ -20,8 +20,7 @@ use rustc_serialize::{
 };
 use rustc_session::Session;
 use rustc_span::hygiene::{
-    ExpnDataDecodeMode, ExpnDataEncodeMode, ExpnId, HygieneDecodeContext, HygieneEncodeContext,
-    SyntaxContext, SyntaxContextData,
+    ExpnId, HygieneDecodeContext, HygieneEncodeContext, SyntaxContext, SyntaxContextData,
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ty/query/on_disk_cache.rs:23]: Nit: could you avoid multi-line and nested imports?

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 3]
```diff
@@ -176,12 +182,12 @@ impl<'a, 'tcx> Encodable<EncodeContext<'a, 'tcx>> for SyntaxContext {
...
 impl<'a, 'tcx> Encodable<EncodeContext<'a, 'tcx>> for ExpnId {
     fn encode(&self, s: &mut EncodeContext<'a, 'tcx>) -> opaque::EncodeResult {
-        rustc_span::hygiene::raw_encode_expn_id(
-            *self,
-            &s.hygiene_ctxt,
-            ExpnDataEncodeMode::Metadata,
-            s,
-        )
+        if self.krate == LOCAL_CRATE {
+            // We will only write details for local expansions.
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:186]: That's true, but why?
What happens if we have a foreign expansion here and don't write details for it?

File: compiler/rustc_span/src/hygiene.rs [hunk 4]
```diff
@@ -1020,11 +1137,18 @@ pub struct HygieneEncodeContext {
 }
 
 impl HygieneEncodeContext {
+    /// Record the fact that we need to serialize the corresponding `ExpnData`.
+    pub fn schedule_for_encoding(&self, expn: ExpnId) {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1141]:
```suggestion
    pub fn schedule_expn_data_for_encoding(&self, expn: ExpnId) {
```
I misinterpreted what this function does, the `ExpnId` itself is encoded immediately, it's `ExpnData` encoding which is delayed.

**Comment by petrochenkov** [comment]
One remaining question about duplicate tables - https://github.com/rust-lang/rust/pull/86676#discussion_r669794333.
Although I now see that the local table's size is used for assigning fresh local IDs at least.

**Comment by cjgillot** [comment]
> One remaining question about duplicate tables - [#86676 (comment)](https://github.com/rust-lang/rust/pull/86676#discussion_r669794333).
> Although I now see that the local table's size is used for assigning fresh local IDs at least.

I have two reasons for keeping separate foreign tables:
- they are expected to be sparse, where the local table is full;
- they should eventually be removed, to fetch the information directly from the metadata decoder like DefIds do.

If these are satisfactory, I will add them as comments.

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 3 above)
Comment (cjgillot) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:186]: In that case, the decoder will fetch the expansion from the other crate's metadata.

Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:186]: Aha, finally found that issue - https://github.com/rust-lang/rust/issues/73047 (and https://github.com/rust-lang/rust/issues/74731).

Sometimes it's preferable for the "other crate" to not exist in its `.rlib` form, proc macro crates and other "macro only" crates clearly "overstay their welcome" when they exist only for the spans to be decodable from them.
I also know that right now we simply drop some spans from proc macro (dependency?) crates so their rlibs are not required, but ideally they would be preserved.

So eventually we'll need some span metadata inlining scheme (which includes hygiene data).

Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:186]: Could you add a comment saying that here we are relying on the other crate's metadata existing, and maybe linking to these issues?

**Comment by petrochenkov** [comment]
r=me after squashing all the review stuff into main commits.

File: compiler/rustc_hir/src/definitions.rs [hunk 5]
```diff
@@ -409,7 +409,7 @@ impl Definitions {
     }
 
     pub fn expansion_that_defined(&self, id: LocalDefId) -> ExpnId {
-        self.expansions_that_defined.get(&id).copied().unwrap_or_else(ExpnId::root)
+        self.expansions_that_defined.get(&id).copied().unwrap_or(ExpnId::root())
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/definitions.rs:412]: Nit: unnecessary change.
